File: 001-020/019.js
//うるう年を求める
let isLeapYear = function (year) {
  if (year % 400 === 0 || (year % 100 !== 0 && year % 4 === 0)) {
    return true;
  } else {
    return false;
  }
};

//2月の最終日を求める
let febDays = function (year) {
  return isLeapYear(year) ? 29 : 28;
};

//各月の最終日を求める
let monthDays = function (month, year) {
  if (month === 2) {
    return febDays(year);
  } else if (month === 4 || month === 6 || month === 9 || month === 11) {
    return 30;
  } else {
    return 31;
  }
}

//初期の曜日として1900年12月1日の曜日を求める
// 0:Monday 1:Tuesday 2:Wednesday 3:Thursday 4:Friday 5:Saturday 6:Sunday
let dayWeek = (4 * 30 + 6 * 31 + febDays(1900)) % 7;
let sundays = 0;
//1901~2000までループ
for (let year = 1901; year <= 2000; year++) {
  //前の月の12月~今月の11月の最終日を求めて翌月の最初の曜日を求める
  for (let month = 0; month <= 11; month++) {
    let tmpM;
    let tmpY;
    if (month === 0) {
      tmpM = 12;
      tmpY = year - 1;
    } else {
      tmpM = month;
      tmpY = year;
    }
    dayWeek += monthDays(tmpM, tmpY) % 7;
    if (dayWeek > 6) {
      dayWeek -= 7;
    }
    if (dayWeek === 6) {
      sundays++;
    }
  }
}

console.log(sundays);

File: 011-020/019.rb
# ruby特有のループ、条件分岐の書き方に書き換えてあるが、
# 解を求めるアルゴリズムは、javascriptのコードと同じ
def feb_days(year)
  if (year % 400).zero? || (year % 100 != 0 && (year % 4).zero?)
    29
  else
    28
  end
end

def month_days(month, year)
  if month == 2
    feb_days(year)
  elsif [4, 6, 9, 11].include?(month)
    30
  else
    31
  end
end

day_week = (4 * 30 + 6 * 31 + feb_days(1900)) % 7
sundays = 0

1901.upto(2000) do |year|
  12.times do |month|
    if month.zero?
      tmp_m = 12
      tmp_y = year - 1
    else
      tmp_m = month
      tmp_y = year
    end
    day_week += month_days(tmp_m, tmp_y) % 7
    day_week -= 7 if day_week > 6
    sundays += 1 if day_week == 6
  end
end

puts sundays
